coremltools/models/utils.py

```diff
@@ -1675,3 +1681,394 @@ def randomize_weights(mlmodel: "_ct.models.MLModel"):
+
+    The implementation of this util is mainly migrated from the below source:
+    https://github.com/apple/ml-stable-diffusion/blob/main/python_coreml_stable_diffusion/chunk_mlprogram.py
+
+    Parameters
+    ----------
+    model_path: str
+        Path to the mlpackage file to be split into two mlpackages of approximately same file size.
+
+    output_dir: str
+        Path to output directory where the two model chunks / pipeline model would be saved.
+        If the `model_path` is `{path}/{model_name}.mlpackage`, the chunks models are going to be saved as:
+
+        1. first chunk model: `{output_dir}/{model_name}_chunk1.mlpackage`
+        2. second chunk model: `{output_dir}/{model_name}_chunk2.mlpackage`
+        3. chunked pipeline model: `{output_dir}/{model_name}_chunked_pipeline.mlpackage`
+
+    remove_original: bool
+        If True, removes the original (non-chunked) model to avoid duplicating storage.
+
+    merge_chunks_to_pipeline: bool
+        If True, model chunks are managed inside a single pipeline model for easier asset maintenance.
+
+    check_output_correctness: bool
+        - If True, compares the outputs of original Core ML model with that of pipelined CoreML model chunks and reports PSNR in dB.
+        - Enabling this feature uses more memory. Disable it if your machine runs out of memory.
+
+    Examples
+    --------
+    .. sourcecode:: python
+
+        import coremltools as ct
+
+        model = ct.models.MLModel("my_model.mlpackage")
+        output_dir = "./output/"
+
+        # The following code will produce two chunks models:
+        # `./output/my_model_chunk1.mlpackage` and `./output/my_model_chunk2.mlpackage`
+        ct.models.utils.bisect_model(
+            model,
+            output_dir,
+        )
+
+        # The following code will produce a single pipeline model `./output/my_model_chunked_pipeline.mlpackage`
+        ct.models.utils.bisect_model(
+            model,
+            output_dir,
+            merge_chunks_to_pipeline=True,
+        )
+    """
+    # We do the lazy import to prevent circular import
+    from coremltools.converters.mil.converter import mil_convert as _mil_convert
+
+    def get_pymil_prog_and_spec_from_model_path(model_path: str):
+        spec = load_spec(model_path)
+
+        # convert the model spec into pymil program,
+        # we also convert operations into type of List
+        prog = _milproto_to_pymil.load(
+            spec,
+            spec.specificationVersion,
+            _try_get_weights_dir_path(model_path),
+        )
+        if len(prog.functions) > 1 or "main" not in prog.functions:
+            raise ValueError("'bisect_model' only support model with a single 'main' function.")
+        
+        func = prog.functions["main"]
+        func.operations = list(func.operations)
+
+        return prog, spec
+
+    # The below implementation assumes that the model is single function, with a "main" function.
```

Comment:
Shall we add error handling here?

Reply:
The handling is here (https://github.com/apple/coremltools/pull/2286/files#diff-e401ca9a79a05199664b882927b7fa56d795167e9b0fb5e5630364910a7a6944R1755).
The corresponding test is here (https://github.com/apple/coremltools/pull/2286/files#diff-d0c9ed9748b0a1bf3bb9b7446f15c9322e9cfae27af41cf4381b16d643ac79f8R947).